### ISSUE TITLE
Is the declaration "Foo<@NullnessUnspecified Object>" a possible mismatch when Foo forbids nullable type arguments?

### ISSUE THREAD

**Issue by cpovirk** [jspecify/jspecify#158]
The premise:

```java
@NullExplicit
interface Foo<T> {}

...
Foo<@NullnessUnspecified Object> foo; // the declaration in question
```

We already have a special rule related to this case:

> [If U is `[@NonNull]`, then U is the type variable usage’s effective nullability specifier](https://docs.google.com/document/d/15NND5nBxMkZ-Us6wz3Pfbt4ODIaWaJ6JDs4w6h9kUaY/edit#heading=h.u8phmghg96wr)

More concretely, this is "the `ImmutableList.Builder` rule": If you have an `ImmutableList.Builder<@NullnessUnspecified Object>`, we _never_ let you put nullable objects into it, even in "lenient mode." In [sample-syntax terms](https://github.com/jspecify/jspecify/blob/9da93c51a4d3c8bb87a10f42eb45dbce091c8a7e/samples/README.md#syntax), we'd call it an outright "mismatch." This is in contrast to a case of "not enough information," as you would have when putting nullable objects into a `List<@NullnessUnspecified Object>`. The difference is that `ImmutableList.Builder` is known to forbid nullable type arguments, while `List` is known to permit them.

What that rule does _not_ speak to is the _declaration_ of the variable/parameter/whatever with type `ImmutableList.Builder<@NullnessUnspecified Object>` in the first place. (Of course it's very unlikely ([perhaps impossible!](https://github.com/jspecify/jspecify/issues/137)) for a user to literally write the declaration in that form. But that type would be the _effective_ type of the object if it's declared in non-nullness-explicit code. And that _will_ happen a lot.)

And the question about the declaration is: Is it a case of "not enough information," or is it completely in the clear?

My starting point is "not enough information." I get there from naive model that that means "mismatch if you replace `@NullnessUnspecified` with `@Nullable`." (And in fact that's what the prototype checker I've been playing with currently does.)

But I see that that's not useful: Because of the special rule discussed above, `@NullnessUnspecified U` doesn't end up letting you do anything that plain `U` wouldn't already. So we already have enough information to ensure that there is no mismatch.

Thus, it would be nice to say that we're completely in the clear. To do so, I think we'd end up needing an additional special rule for checking validity of parameterized types. That's a bit different than our usual policy, which has been to say "just apply normal JLS rules \[here, checking that the arg is a subtype of the bound\] but extended to augmented types." Still, we could write such a rule if we wanted.

That said: I've experimented with running an annotated Guava through the prototype checker in "strict mode" (meaning that "not enough information" triggers an error). In lenient mode (meaning that "not enough information" findings are ignored), Guava passes. (OK, there are currently something like a dozen failures that I haven't dealt with, but it's all minor stuff.) In strict mode, IIRC, there are several hundred errors.

So: As a practical matter, I don't think anyone is going to try to use strict mode unless the plan is to fully stamp out unspecified nullness. If that's the case, then I'm not sure it's worth the effort for us to write a special case (and tool authors to implement it) that applies only in the case of strict mode + unspecified nullness (since there will be no problem reported in lenient mode regardless).

I'm not dead set against such a rule; I just think that the existing special case ("`ImmutableList.Builder` rule") is more important than this one. (And again, the existing rule fits fine with a policy of "just apply normal JLS rules but extended to augmented types.")

(Note that are existing samples that annotate lines "like this" with `jspecify_nullness_not_enough_information` -- like [in `UnspecifiedTypeArgumentForNonNullableParameter`](https://github.com/jspecify/jspecify/blob/9da93c51a4d3c8bb87a10f42eb45dbce091c8a7e/samples/UnspecifiedTypeArgumentForNonNullableParameter.java#L28). If we change course here, we'll want to update them.)

([context](https://github.com/jspecify/jspecify/pull/156/commits/b41f58dbb4fd352b4bbf4eae591468c3b51ee7ba#diff-70f0b627440fb42f5d38532e4c3f8713e67d146d308a45b570d9a2107e2aa8cbR52))

**Comment by kevin1e100** [comment]
Right so I wrote the `ImmutableList.Builder` rule a long time ago and I still think it's valuable. _However_ I think that it was mostly written to handle the case of `ImmutableList<String>` in unspecified context: by "normal" rules, we'd consider `String` here to be implicitly of unspecified nullness, but that seems unfortunate.  So, by taking `ImmutableList`'s type parameter bound into account, we get a list of non-null Strings _even in unspecified context_.  So far so good.

But, if someone goes through the trouble of _explicitly_ writing `ImmutableList.Builder<@NullnessUnspecified String>`, the above rule alone would just ignore the explicit annotation, which seems problematic.

Note that Guava's `ImmutableList.Builder` will just flat-out throw when presented with `null` values.  That both means that specifying `ImmutableList.Builder<@NullnessUnspecified String>` to allow unspecified Strings is ultimately "harmless" in some sense, since it'll throw regardless, but also pointless, since adding `null` really will be an error.  Plus, if we're in an unspecified context then values presented will almost certainly be of unspecified nullness, which checkers will generally allow.

So for `ImmutableList.Builder` it seems fine and even useful to ignore an explicit `ImmutableList.Builder<@NullnessUnspecified String>`.  I'm less clear whether that's the right choice _in general_.  Either way, the point is mute if we don't introduce `@NullnessUnspecified` as an explicit type annotation.

**Comment by kevinb9n** [comment]
Naive viewpoint, almost entirely innocent of the latest thinking on unspec. This is just fantasy of how I think I would _like_ it to work; please tell me what I'm missing.

A NonNull-bounded type parameter like in `@NullMarked class Foo<T>` tells us something quite definitive: the class supports only ONE definitive policy on null, for all usages. For all unprojected `T` values in its API, as out-types it will never produce `null` and as in-types it has the right to reject `null` (or even just behave very badly on it).

So first off, I would not consider the type `Foo<@Nullable String>` to even exist (it would be wrong to write that in either context).

As for `Foo<@Unspec String>`, my own crappy form of the many-worlds model tells me: there's only one world here! That is, it becomes functionally identical to `Foo<@NonNull String>`. The argument type has specified nullness even though _I_ didn't specify it myself!

So in null-unmarked code, `Foo<String>` is implicitly short for `Foo<@NonNull String>` (that annotation being officially redundant).

And assuming yes on #137, then writing `Foo<@Unspec String>` explicitly in _either_ context should be flagged somehow because it can't possibly be the way to say what you really mean.

Stepping back and thinking about the migration curve: on Foo's out-types we simply gain information; on its in-types we may have a new source of errors, but that seems appropriate (these would only be cases of passing outright `@Nullable` references in, unless they're using the strict mode we don't expect people to use, and even that is too much trouble, why have they turned on null analysis yet in the first place?)

**Comment by kevinb9n** [comment]
I notice that the namesake of the "`ImmutableList.Builder` rule" is a consume-only API. Any particular reason for that? Are we equally willing to assume the elements coming out of an `ImmutableList<Foo>` are non-null (when analyzing unmarked code)? 

(Or is it that the difference just doesn't matter much, since we expect tools to do lenient type conversions anyway.)

**Comment by cpovirk** [comment]
Sorry, this one of many threads that I already owed a response on, but I can answer that one.

Yes, it equally applies to elements coming out. This can matter for cases like `FluentIterable.first()` and `FluentIterable.toList()` ([previously](https://github.com/jspecify/jspecify/issues/157#issuecomment-1030292196)), whose return type is `Optional<E>` or `ImmutableList<E>` for a possibly nullable type `E`. It's nice for checkers to still know that calling `get` on the returned objects will never produce `null`.

**Comment by kevinb9n** [comment]
In that case I believe we should enforce that those methods return e.g. `Optional<@NonNull E>`, as argued on #86. 

It still seems like the right rule to have, but more for reasons of my item (3) above?

**Comment by cpovirk** [comment]
OK, RE: [(3)](https://github.com/jspecify/jspecify/issues/158#issuecomment-1032057973): I would guess that there are many cases in which an unspecified type "could mean only one thing." I would hesitate to make blanket statements about them, but I feel reasonably good about our existing "`ImmutableList.Builder` rule" special case, given the specific problems it catches.

That is: The namesake case of `ImmutableList.Builder<@Nullable String>` is a case in we'd be happy for tools to put up compile-time obstacles to getting a null value in: Those obstacles protect against runtime errors. (And even if tools insert _runtime_ checks, those checks would throw an exception only when one was already going to be thrown anyway.)

But you could imagine other cases: Given a supermethod that returns `String`, a submethod "could only mean" to return `String`, too -- even if it declares a return type of `@Nullable String` because it really does return `null` in some cases, thanks to a regrettable programming decision from years ago. But I think we've generally felt that we want for a tool to be able to check callers by looking only at the submethod's annotations, even though it could "know" more if it looked at the supermethod, too.

I think it's fine for us to be "inconsistent" about this: It can probably work fine for us to have a special rule for substitution (the "`ImmutableList.Builder` rule") but not have special rules for, e.g., return types.

I see it as a cost-benefit calculation. Conveniently, the substitution rule requires relatively little effort in the spec and (I hope) in tools. But ideally, we can focus on the cost-benefit ratio to actual user code: I think that the "`ImmutableList.Builder` rule" is primarily benefit (not necessarily completely -- maybe we should at least go over that?), while the hypothetical return-type rule is probably primarily cost.

**Comment by kevinb9n** [comment]
My thinking on this has evolved and I owe a new summary.

**Comment by cpovirk** [comment]
A kevinb9n point from a few months ago popped into my head: When we're missing information, we might be more interested in inferring what the information _should_ be than we are in directly catching bugs.

Under that model, I'm wondering if it's yet more important that the scenario under discussion be a possible mismatch.

The idea would be this: `Foo` is some unannotated common library with some unannotated users. An automated tool might say: Hey, let's try to add some nullness information to `Foo`, starting with its type parameter's bound. So it might ask:
- If I set the bound to `@Nullable Object`, does that introduce any possible mismatches?
- If I set the bound to `@NonNull Object`, does that introduce any possible mismatches?

I'm sure there will be a number of cases in which the answer to _both_ question is "yes" or in which the answer to _both_ questions is "no." But sometimes, the answer will be that one annotation introduces possible mismatches and the other does not. In those cases, the tool might add the annotation that doesn't introduce any possible mismatches.

For that purpose, we might find it useful to aggressively flag possible mismatches. For example, suppose that `Consumer` is unannotated, and suppose that a tool wants to experiment with a `@NonNull Object` bound for it. When the tool scans unannotated user code and encounters a `Consumer<Object>`, it may be useful for that tool to notice that _we don't know_ if `extends @NonNull Object` would be a problem for that user. Adding `extends @NonNull Object` would create a _possible mismatch_.

In short: Yes, when we _have_ information about a type parameter, then we can let unspecified-nullness arguments slide. (Tools accomplish that by not warning/error-ing about "possible mismatches" at all.) But when we _don't_ have information about a type parameter, it seems useful to be able to ask the question "If we _guessed_ that the type was _X_, is that known to be OK, or would it force us to let _more_ things slide?"

**Comment by kevinb9n** [comment]
The working decision is to do it the way the reference checker is currently doing it, but I have to leave it to @cpovirk to describe that.

**Comment by cpovirk** [comment]
The description of the current reference checker behavior is easy, at least: If `Foo` forbids nullable type arguments, then the checker says that `Foo<Object*>` _is_ a possible mismatch.

(Reminder: In practice, we expect that most tools will not warn users about possible mismatches, including cases like dereferencing a value that has unspecified nullness. Notably, that is the case for the current reference checker in its default mode. Essentially, tools will conclude "There's a _possible_ mismatch here, but we aren't concerned about possible mismatches." From the user perspective, this isn't much different than saying that there's no possible mismatch at all.)

That answer of "possible mismatch" falls naturally out of our "many worlds" [model](https://github.com/jspecify/jspecify/wiki/nullness-unspecified#behavior-may-vary) for unspecified nullness: The type argument has unspecified nullness, so the checker asks whether both of the possible specifications for that argument (namely, `Foo<Object!>` and `Foo<Object?>`) would be allowed. It turns out that `Foo<Object!>` would be allowed, while `Foo<Object?>` would not, so we have a possible mismatch.

**Comment by cpovirk** [comment]
For 1.0 purposes, then, I'll lay out the argument _against_ calling this a possible mismatch. Based on what I said in my previous post, I think it's fair to characterize that argument as an argument for a _special case_: If we look at [the example from above](https://github.com/jspecify/jspecify/issues/158#issuecomment-1058374368) again, we seem to agree that dereferencing a value that has unspecified nullness is a "possible mismatch" (as well as possibly a hint that the value's type is actually non-null). But we've been disagreeing on this issue's particular case of type arguments.

I see 3 main reasons for arguing for that special case:

1. It might fit in with a model under which types with out-of-bounds type arguments like `Foo<Object?>` "do not exist" (and even `Foo<Object*>` "does not exist") because they're automatically "replaced" by a similar type with in-bounds type arguments, `Foo<Object!>`. We've touched on that model a couple times, and I have [concerns](https://github.com/jspecify/jspecify/issues/86#issuecomment-1058471961). But for purposes of this issue, I'd like to figure out whether it's actually orthogonal: It seems like we _could_ say that `Foo<Object*>` is a potential mismatch and _still_ choose to treat the type as `Foo<Object!>` from that point forward. And in fact we'd probably "need" to do that so that we can still report `Foo<Object?>` (think "`ImmutableList<@Nullable String>`") as a mismatch and treat it like `Foo<Object!>` afterward.

2. Out-of-bounds type arguments _already_ trigger a special case, one that we've called ["the `ImmutableList.Builder` rule"](https://github.com/jspecify/jspecify/issues/158#issue-756650991). (Reminder: I believe that the "does not exist" model of item 1 above is meant to _replace_ that special case. That's all the more reason to figure out how closely that model is tied to this current issue!) And that may provide more than just "precedent": "The `ImmutableList.Builder` rule" might actually guarantee that almost the all _usages_ of `Foo<Object*>` are treated almost like `Foo<Object!>` already. Maybe that helps us feel less "need" to report a possible mismatch at the type argument, knowing that `Foo<Object?>` would mostly be treated like `Foo<Object!>`, anyway?

3. This case seems like a relatively tractable one. If we believe that such special cases are generally useful, then we might want lots of them, limited only by what we believe would be practical to implement. And maybe it's _hard_ to see a dereference of a local variable somewhere and work backward to figure out which method calls had their results assigned to that variable (and thus which methods must have a non-null return type). But maybe it's _easy_ to loosen our checks for type arguments. Maybe it would be something like this: "If the type-parameter bound is non-null, then instead of checking that the type argument is a subtype of that, check that _the non-null projection of_ the type argument is a subtype of that." (That way, we still check any type arguments inside the type argument! Oh, and maybe we'd want an exception for _nullable_ type arguments so that we still call them mismatches?)

**Comment by kevinb9n** [comment]
Since substantial new elucidation appeared after this was marked "working decision", I'm removing that badge for now.

**Comment by netdpb** [comment]
**Current decision**:

In the case of

```
@NullMarked interface Foo<T extends @NonNull Object> {
  T returnT();
}
```

when tools encounter `@NullUnmarked Foo<String> foo();`, they should treat it as a ""possible mismatch"", because in the world where the user meant `Foo<@Nullable String>` that would represent an out-of-bounds type error.

Tools are free to report such possible mismatches to their users or not. Tools are free to refine the type to `Foo<@NonNull String>` for references to it in implementation code:

```
// test:risky-conversion:String* to String!
Foo<String> foo = ...;

foo.returnT().size(); // tools could decide not to report this
```

**Argument for changing**: In that case, we know that there's only one valid nullability for `Foo<String>`, so there aren't "many worlds" after all. So it's better just to interpret it as if it were `Foo<@NonNull String>` anyway.

**Timing**: This must be decided before version 1.0 of the spec.

**Proposal for 1.0: Finalize the current decision.** If you agree, please add a thumbs-up emoji (👍) to this comment. If you disagree, please add a thumbs-down emoji (👎) to this comment and briefly explain your disagreement. Please only add a thumbs-down if you feel you can make a strong case why this decision will be materially worse for users or tool providers than an alternative.

**Comment by kevinb9n** [comment]
Note there is no working decision here to finalize.

**Comment by netdpb** [comment]
I'm still convinced that an explicit usage of `Foo!<Bar*>` when `Foo`'s type parameter's bound is `SuperBar!` is a clear-cut case of a possible mismatch, more specifically an *out-of-bounds type argument in some world*.

While it's true that there is only one valid world (the one in which the usage is intended to be `Foo!<Bar!>`, not `Foo!<Bar?>`), that is exactly what it means to say that there's an out-of-bounds type argument in some world. If we instead declare that the other worlds *don't exist*, that devalues the entire multiple-worlds framing.

As @cpovirk said, saying that `Foo!<Bar*>` is a possible mismatch doesn't stop tools from treating expressions that use the symbol declared like that as if they were `Foo!<Bar!>` using a kind of optimistic flow-typing. JSpecify has always said that conformant tools can use any extra information they have to refine their nullness checking, and in fact can report anything to their users.

**Comment by cpovirk** [comment]
This came up in a meeting with @netdpb and @wmdietl. Here is my position, which has evolved slightly in the succeeding 45 minutes:

- As a further conceptual justification for [the earlier proposal](https://github.com/jspecify/jspecify/issues/158#issuecomment-2045971747), I like [netdpb's Wednesday framing](https://github.com/jspecify/jspecify/issues/158#issuecomment-2344786796) a lot.
- As a practical justification for that proposal, I still like [applying kevinb9n's point about inferring missing information](https://github.com/jspecify/jspecify/issues/158#issuecomment-1252699448).
- In a narrow sense, there is currently nothing _in the spec_ that speaks to the question as asked in the issue title: The spec defines rules for subtyping, etc., and our main answer for "What do checkers need to know to do with that information?" has been the conformance tests. That suggests that this issue is not in fact a spec-1.0 blocker, and I'm going to tentatively remove it from that milestone (after initially choosing to carry it over from our earlier list).
- However, it is still very possible to argue that this _is_ related to spec 1.0 because the spec _could_ define the rule [mentioned above](https://github.com/jspecify/jspecify/issues/158#issuecomment-1032057973) under which [`Optional<String?>` "does not exist" and therefore `Optional<String*>` means exactly `Optional<String!>`](https://github.com/jspecify/jspecify/issues/640). My claim is on [#640](https://github.com/jspecify/jspecify/issues/640) is that that should be a post-1.0 discussion, since it is mostly a "red-to-green" change for users (with an effect for users of strict checkers run on unannotated code only). There could then be some further discussion over whether the policy "`Optional<String*>` means exactly `Optional<String!>`" means "And checkers report that they discarded the possible mismatch in a different way than they normally report a possible mismatch (or not at all)." That would bring us back to the question stated in the issue title and to the associated conceptual and practical arguments.

**Comment by cpovirk** [comment]
Right, sorry, only the latter point. https://jspecify.dev/docs/spec/#the-scope-of-this-spec tries to disclaim all that while at the same time gesturing suggestively in this direction. I think I would personally be open to getting into things like that someday, but I have been reluctant to do so as a blocker for 1.0.
